src/Handlers/AbstractPaymentMethodHandler.php

```diff
@@ -0,0 +1,684 @@
+        } catch (Exception $exception) {
+            $message = sprintf(
+                "There was an error with the payment method. Order number: %s Missing data: %s",
+                $transaction->getOrder()->getOrderNumber(),
+                $exception->getMessage()
+            );
+            $this->logger->error($message);
+            throw new AsyncPaymentProcessException($transactionId, $message);
+        }
+
+        try {
+            $response = $this->checkoutService->payments($request);
+        } catch (AdyenException $exception) {
+            $message = sprintf(
+                "There was an error with the /payments request. Order number %s: %s",
+                $transaction->getOrder()->getOrderNumber(),
+                $exception->getMessage()
+            );
+
+            $this->logger->error($message);
+
+            throw new AsyncPaymentProcessException($transactionId, $message);
+        }
+
+        $orderNumber = $transaction->getOrder()->getOrderNumber();
+
+        if (empty($orderNumber)) {
+            $message = 'Order number is missing';
+            throw new AsyncPaymentProcessException($transactionId, $message);
+        }
+
+        $result = $this->paymentResponseHandler->handlePaymentResponse($response, $orderNumber, $salesChannelContext);
+
+        try {
+            $this->paymentResponseHandler->handleShopwareApis($transaction, $salesChannelContext, $result);
+        } catch (PaymentCancelledException $exception) {
+            throw new CustomerCanceledAsyncPaymentException($transactionId, $exception->getMessage());
+        } catch (PaymentFailedException $exception) {
+            throw new AsyncPaymentFinalizeException($transactionId, $exception->getMessage());
+        }
+
+        // Payment had no error, continue the process
+        return new RedirectResponse($transaction->getReturnUrl());
+    }
+
+    /**
+     * @param AsyncPaymentTransactionStruct $transaction
+     * @param Request $request
+     * @param SalesChannelContext $salesChannelContext
+     */
+    public function finalize(
+        AsyncPaymentTransactionStruct $transaction,
+        Request $request,
+        SalesChannelContext $salesChannelContext
+    ): void {
+        $transactionId = $transaction->getOrderTransaction()->getId();
+        $this->checkoutService->startClient($salesChannelContext->getSalesChannel()->getId());
+        try {
+            $this->resultHandler->processResult($transaction, $request, $salesChannelContext);
+        } catch (PaymentCancelledException $exception) {
+            throw new CustomerCanceledAsyncPaymentException($transactionId, $exception->getMessage());
+        } catch (PaymentFailedException $exception) {
+            throw new AsyncPaymentFinalizeException($transactionId, $exception->getMessage());
+        }
+    }
+
+    /**
+     * @param string $address
+     * @return array
+     */
+    public function splitStreetAddressHouseNumber(string $address): array
```

Comment:
let's create a ticket to move the following functions to a helper or builder class(es)

Reply:
PW-3119